lib/pbench/test/unit/agent/task/test_tools_kill.py

```diff
@@ -0,0 +1,440 @@
+                self.file_name = file_name
+                self.display_name = display_name
+
+            def load(self, tm_dir: pathlib.Path, uuid: str) -> bool:
+                action_list.append(("load", self.file_name, str(tm_dir), uuid))
+                return True
+
+            def killem(self, echo: Callable[[str], None]) -> None:
+                action_list.append(("killem", self.display_name))
+
+        def mock_gen_run_directories(
+            pbench_run_dir: pathlib.Path,
+        ) -> Iterable[Tuple[pathlib.Path, str]]:
+            for run_dir in [
+                (pathlib.Path("run0/tm"), "uuid1abc"),
+                (pathlib.Path("run1/tm"), "uuid2def"),
+                (pathlib.Path("run2/tm"), "uuid3ghi"),
+            ]:
+                yield run_dir
+
+        def mock_gen_host_names(run_dir: pathlib.Path) -> str:
+            hosts = {"run0": ["hostA", "hostB"], "run2": ["host3"]}
+            for host in hosts.get(run_dir.name, []):
+                yield host
+
+        class MockTemplateSsh:
+            def __init__(self, name: str, ssh_opts: List[str], cmd: str):
+                self.name = name
+                self.ssh_opts = ssh_opts
+                self.cmd = cmd
+
+            def start(self, host: str, **kwargs):
+                action_list.append(("start", host, repr(kwargs)))
+
+            def wait(self, host: str):
+                action_list.append(("wait", host))
+
+        runner = CliRunner(mix_stderr=False)
+        monkeypatch.setattr(
+            "pbench.cli.agent.commands.tools.kill.PidSource", MockPidSource
+        )
+        monkeypatch.setattr(
+            "pbench.cli.agent.commands.tools.kill.gen_run_directories",
+            mock_gen_run_directories,
+        )
+        monkeypatch.setattr(
+            "pbench.cli.agent.commands.tools.kill.gen_host_names",
+            mock_gen_host_names,
+        )
+        monkeypatch.setattr(
+            "pbench.cli.agent.commands.tools.kill.TemplateSsh",
+            MockTemplateSsh,
+        )
+        result = runner.invoke(kill.main, catch_exceptions=False)
+        assert (
+            result.exit_code == 0
+        ), f"stdout={result.stdout!r}, stderr={result.stderr!r}"
+        expected_action_list = [
+            ("load", "redis.pid", "run0/tm", "uuid1abc"),
+            ("load", "pbench-tool-data-sink.pid", "run0/tm", "uuid1abc"),
+            ("load", "tm.pid", "run0/tm", "uuid1abc"),
+            ("load", "redis.pid", "run1/tm", "uuid2def"),
+            ("load", "pbench-tool-data-sink.pid", "run1/tm", "uuid2def"),
+            ("load", "tm.pid", "run1/tm", "uuid2def"),
+            ("load", "redis.pid", "run2/tm", "uuid3ghi"),
+            ("load", "pbench-tool-data-sink.pid", "run2/tm", "uuid3ghi"),
+            ("load", "tm.pid", "run2/tm", "uuid3ghi"),
+            ("killem", "redis server"),
+            ("killem", "tool data sink"),
+            ("killem", "local tool meister"),
```

Comment:
Why do we load nine and kill only three?

Reply:
Because `killem()` is `kill()`ing all the PIDs for each of the PID sources, Redis, TDS, and TM.  We have 3 run directories, with 3 PID sources, for a total of 9 loads.